### PR TITLE
base64 and base58 crates benchmarks

### PR DESCRIPTION
**Pull request by kchalkias**
Issue: https://github.com/MystenLabs/fastcrypto/issues/168 and https://github.com/MystenLabs/fastcrypto/issues/169

**TL;DR for base64**
- `radix64` using vectorized instructions is the fastest.
- `base64ct` is constant time and non-malleable in decoding, but as the message grows it becomes more expensive (for >64B messages it's 4x more expensive than the rest). 
- PS: Also had a measurement on 320bytes input, and `base64ct` was 4x slower than `radix64`, and `radix64` was 15% faster than `base64` on decoding, so that's probably the amortized diff factor.

**Recommendations:**
- **for Encoding**: for critical confidential data (i.e., private keys) prefer `base64ct`, else `radix64`.
- **for Decoding**: for critical confidential data (i.e., private keys) or when malleability is important prefer `base64ct`, else `radix64`.

**TL;DR for base58**
- `bs58` is faster for encoding for up to 32B, then `base58` wins, `rust-base58` is too slow (don't use it).
- `base58` is faster in decoding after 20B. For 64B it's 3x faster than `bs58` and 10x than `rust-base58`.
- All apis are not constant time (as we notice from the different results on ZERO vs RANDOM msgs). This is expected due to how base58 works. Performance is also deteriorating really fast as the msgs size grows (also expected).
- Base58 is many times more expensive than base64 (90x slower in encoding 64B payloads, 16x slower in decoding)

_@joyqvq ^^ this wasn't expected, I'd expect `bs58` to be faster according to their authors, unless I need some flag to enable optimizations._ 

**Recommendations:**
- **for Encoding**: for <=32B use `bs58`, for longer msgs use `base58`
- **for Decoding**: for <=20B use `bs58`, for longer msgs use `base58`

**--benchmarks for base64 (Macbook Pro M1 Max)--**
```
Benchmarking base64_encode_MSG_20B_RANDOM
                        time:   [45.915 ns 45.959 ns 46.008 ns]

Benchmarking base64ct_encode_MSG_20B_RANDOM
                        time:   [45.026 ns 45.121 ns 45.236 ns]

Benchmarking radix64_encode_MSG_20B_RANDOM
                        time:   [37.804 ns 37.826 ns 37.852 ns]

Benchmarking base64_encode_MSG_32B_RANDOM
                        time:   [48.011 ns 48.152 ns 48.350 ns]

Benchmarking base64ct_encode_MSG_32B_RANDOM
                        time:   [52.572 ns 52.630 ns 52.693 ns]

Benchmarking radix64_encode_MSG_32B_RANDOM
                        time:   [39.287 ns 39.327 ns 39.371 ns]

Benchmarking base64_encode_MSG_64B_RANDOM
                        time:   [57.762 ns 57.788 ns 57.819 ns]

Benchmarking base64ct_encode_MSG_64B_RANDOM
                        time:   [70.838 ns 70.911 ns 70.996 ns]

Benchmarking radix64_encode_MSG_64B_RANDOM
                        time:   [49.109 ns 49.166 ns 49.244 ns]

---

Benchmarking base64_decode_MSG_20B_RANDOM
                        time:   [78.911 ns 78.995 ns 79.089 ns]

Benchmarking base64ct_decode_MSG_20B_RANDOM
                        time:   [93.691 ns 93.899 ns 94.130 ns]

Benchmarking radix64_decode_MSG_20B_RANDOM
                        time:   [43.018 ns 43.052 ns 43.094 ns]

Benchmarking base64_decode_MSG_32B_RANDOM
                        time:   [81.408 ns 81.501 ns 81.594 ns]

Benchmarking base64ct_decode_MSG_32B_RANDOM
                        time:   [121.53 ns 121.67 ns 121.83 ns]

Benchmarking radix64_decode_MSG_32B_RANDOM
                        time:   [43.994 ns 44.020 ns 44.051 ns]

Benchmarking base64_decode_MSG_64B_RANDOM
                        time:   [59.192 ns 59.244 ns 59.315 ns]

Benchmarking base64ct_decode_MSG_64B_RANDOM
                        time:   [199.34 ns 199.84 ns 200.55 ns]

Benchmarking radix64_decode_MSG_64B_RANDOM
                        time:   [56.404 ns 56.453 ns 56.509 ns]
```

**--benchmarks for base58 (Macbook Pro M1 Max)--**
```
bs58_encode_MSG_20B_ZERO
                        time:   [69.759 ns 69.901 ns 70.095 ns]

base58_encode_MSG_20B_ZERO
                        time:   [115.55 ns 115.67 ns 115.81 ns]

rust_base58_encode_MSG_20B_ZERO
                        time:   [198.20 ns 198.31 ns 198.45 ns]

bs58_encode_MSG_32B_ZERO
                        time:   [91.042 ns 91.415 ns 91.790 ns]

base58_encode_MSG_32B_ZERO
                        time:   [125.15 ns 125.23 ns 125.33 ns]

rust_base58_encode_MSG_32B_ZERO
                        time:   [227.63 ns 227.96 ns 228.28 ns]

bs58_encode_MSG_64B_ZERO
                        time:   [149.57 ns 149.61 ns 149.66 ns]

base58_encode_MSG_64B_ZERO
                        time:   [181.97 ns 182.15 ns 182.41 ns]

rust_base58_encode_MSG_64B_ZERO
                        time:   [348.41 ns 348.58 ns 348.79 ns]

bs58_encode_MSG_20B_RANDOM
                        time:   [604.39 ns 605.11 ns 606.06 ns]

base58_encode_MSG_20B_RANDOM
                        time:   [655.31 ns 656.53 ns 657.84 ns]

rust_base58_encode_MSG_20B_RANDOM
                        time:   [2.6206 µs 2.6267 µs 2.6342 µs]

bs58_encode_MSG_32B_RANDOM
                        time:   [1.2817 µs 1.2864 µs 1.2944 µs]

base58_encode_MSG_32B_RANDOM
                        time:   [1.3257 µs 1.3272 µs 1.3287 µs]

rust_base58_encode_MSG_32B_RANDOM
                        time:   [4.3458 µs 4.3604 µs 4.3792 µs]

bs58_encode_MSG_64B_RANDOM
                        time:   [4.9817 µs 4.9885 µs 4.9957 µs]

base58_encode_MSG_64B_RANDOM
                        time:   [4.6285 µs 4.6305 µs 4.6329 µs]

rust_base58_encode_MSG_64B_RANDOM
                        time:   [11.088 µs 11.101 µs 11.114 µs]

---

bs58_decode_MSG_20B_ZERO
                        time:   [58.152 ns 58.211 ns 58.292 ns]

base58_decode_MSG_20B_ZERO
                        time:   [136.01 ns 136.51 ns 136.99 ns]

rust_base58_decode_MSG_20B_ZERO
                        time:   [751.09 ns 752.37 ns 753.80 ns]

bs58_decode_MSG_32B_ZERO
                        time:   [77.033 ns 77.067 ns 77.114 ns]

base58_decode_MSG_32B_ZERO
                        time:   [140.64 ns 140.88 ns 141.19 ns]

rust_base58_decode_MSG_32B_ZERO
                        time:   [1.1983 µs 1.1993 µs 1.2004 µs]

bs58_decode_MSG_64B_ZERO
                        time:   [131.98 ns 132.10 ns 132.25 ns]

base58_decode_MSG_64B_ZERO
                        time:   [166.63 ns 166.82 ns 167.06 ns]

rust_base58_decode_MSG_64B_ZERO
                        time:   [2.4537 µs 2.4563 µs 2.4592 µs]

bs58_decode_MSG_20B_RANDOM
                        time:   [340.80 ns 340.89 ns 340.99 ns]

base58_decode_MSG_20B_RANDOM
                        time:   [378.46 ns 379.01 ns 379.61 ns]

rust_base58_decode_MSG_20B_RANDOM
                        time:   [2.5045 µs 2.5079 µs 2.5116 µs]

bs58_decode_MSG_32B_RANDOM
                        time:   [708.07 ns 708.37 ns 708.72 ns]

base58_decode_MSG_32B_RANDOM
                        time:   [512.94 ns 513.22 ns 513.53 ns]

rust_base58_decode_MSG_32B_RANDOM
                        time:   [4.1104 µs 4.1155 µs 4.1214 µs]

bs58_decode_MSG_64B_RANDOM
                        time:   [2.5406 µs 2.5593 µs 2.5798 µs]

base58_decode_MSG_64B_RANDOM
                        time:   [886.29 ns 886.73 ns 887.26 ns]

rust_base58_decode_MSG_64B_RANDOM
                        time:   [8.8275 µs 8.8336 µs 8.8401 µs]
```

### CODE REVIEW

**Review by jonas-lj — APPROVED** [review]

File: fastcrypto/benches/encoding.rs [hunk 1]
```diff
@@ -5,6 +5,7 @@ extern crate criterion;
 
 mod encoding_benches {
     use super::*;
+    use base64ct::{Base64, Encoding};
```
Comment (benr-ml) [inline fastcrypto/benches/encoding.rs:8]: maybe 'Base64 as Base64Ct' to make the call site below more explicit?

**Review by benr-ml — CHANGES_REQUESTED** [review]
